// @ionic/angular/dist/providers/nav-controller.js
import * as tslib_1 from "tslib";
import { Location } from '@angular/common';
import { Injectable, Optional } from '@angular/core';
import { NavigationStart, Router } from '@angular/router';
import { Platform } from './platform';
var NavController = /** @class */ (function () {
    function NavController(platform, location, router) {
        var _this = this;
        this.location = location;
        this.router = router;
        this.direction = DEFAULT_DIRECTION;
        this.animated = DEFAULT_ANIMATED;
        this.guessDirection = 'forward';
        this.guessAnimation = false;
        this.lastNavId = -1;
        // Subscribe to router events to detect direction
        if (router) {
            router.events.subscribe(function (ev) {
                if (ev instanceof NavigationStart) {
                    var id = (ev.restoredState) ? ev.restoredState.navigationId : ev.id;
                    _this.guessAnimation = !ev.restoredState;
                    _this.guessDirection = id < _this.lastNavId ? 'back' : 'forward';
                    _this.lastNavId = _this.guessDirection === 'forward' ? ev.id : id;
                }
            });
        }
        // Subscribe to backButton events
        platform.backButton.subscribeWithPriority(0, function () { return _this.goBack(); });
    }
    NavController.prototype.navigateForward = function (url, animated, extras) {
        this.setDirection('forward', animated);
        if (Array.isArray(url)) {
            return this.router.navigate(url, extras);
        }
        else {
            return this.router.navigateByUrl(url, extras);
        }
    };
    NavController.prototype.navigateBack = function (url, animated, extras) {
        this.setDirection('back', animated);
        // extras = { replaceUrl: true, ...extras };
        if (Array.isArray(url)) {
            return this.router.navigate(url, extras);
        }
        else {
            return this.router.navigateByUrl(url, extras);
        }
    };
    NavController.prototype.navigateRoot = function (url, animated, extras) {
        this.setDirection('root', animated);
        if (Array.isArray(url)) {
            return this.router.navigate(url, extras);
        }
        else {
            return this.router.navigateByUrl(url, extras);
        }
    };
    NavController.prototype.goBack = function (animated) {
        this.setDirection('back', animated);
        return this.location.back();
    };
    NavController.prototype.setDirection = function (direction, animated) {
        this.direction = direction;
        this.animated = (animated === undefined)
            ? direction !== 'root'
            : animated;
    };
    NavController.prototype.consumeTransition = function () {
        var direction = 'root';
        var animated = false;
        if (this.direction === 'auto') {
            direction = this.guessDirection;
            animated = this.guessAnimation;
        }
        else {
            animated = this.animated;
            direction = this.direction;
        }
        this.direction = DEFAULT_DIRECTION;
        this.animated = DEFAULT_ANIMATED;
        return {
            direction: direction,
            animated: animated
        };
    };
    NavController = tslib_1.__decorate([
        Injectable(),
        tslib_1.__param(2, Optional()),
        tslib_1.__metadata("design:paramtypes", [Platform,
            Location,
            Router])
    ], NavController);
    return NavController;
}());
export { NavController };
var DEFAULT_DIRECTION = 'auto';
var DEFAULT_ANIMATED = false;
